src/test/java/com/tabnine/unitTests/BinaryProcessRequesterProviderTests.java

```diff
@@ -0,0 +1,173 @@
+    when(binaryProcessGatewayProvider.generateBinaryProcessGateway())
+        .thenReturn(binaryProcessGateway);
+  }
+
+  @AfterEach
+  public void cleanup() {
+    utilsMockedStatic.close();
+  }
+
+  @Test
+  public void shouldNotInitProcessWhenBinaryInitIsNull() throws IOException {
+    binaryProcessRequesterProvider =
+        BinaryProcessRequesterProvider.create(binaryRun, binaryProcessGatewayProvider, 0);
+
+    executeOnDead();
+
+    verify(binaryProcessGateway, timesBeyondCreation(0)).init(any());
+  }
+
+  @Test
+  public void shouldNotInitProcessWhenBinaryInitIsNotDone() throws IOException {
+    when(mockedFuture.isDone()).thenReturn(false);
+    binaryProcessRequesterProvider =
+        BinaryProcessRequesterProvider.create(binaryRun, binaryProcessGatewayProvider, 0);
+
+    executeOnDead();
+
+    verify(binaryProcessGateway, timesBeyondCreation(0)).init(any());
+  }
+
+  @Test
+  public void shouldInitProcessWhenBinaryInitIsDone() throws IOException {
+    when(mockedFuture.isDone()).thenReturn(true);
+    binaryProcessRequesterProvider =
+        BinaryProcessRequesterProvider.create(binaryRun, binaryProcessGatewayProvider, 0);
+
+    executeOnDead();
+
+    verify(binaryProcessGateway, timesBeyondCreation(1)).init(any());
+  }
+
+  @Test
+  public void shouldInitProcessOnceOnMultipleCallsWithinBackoffTime() throws IOException {
+    MockedStatic<StaticConfig> staticConfigMockedStatic = mockExponentialBackoffWith(5000);
+    when(mockedFuture.isDone()).thenReturn(true);
+    binaryProcessRequesterProvider =
+        BinaryProcessRequesterProvider.create(binaryRun, binaryProcessGatewayProvider, 0);
+
+    executeOnDead(10);
+    staticConfigMockedStatic.close();
+
+    verify(binaryProcessGateway, timesBeyondCreation(1)).init(any());
+  }
+
+  @Test
+  public void shouldInitProcessMultipleTimesOnMultipleCallsOutsideBackoffTime() throws IOException {
+    MockedStatic<StaticConfig> staticConfigMockedStatic = mockExponentialBackoffWith(0);
+    when(mockedFuture.isDone()).thenReturn(true);
+    binaryProcessRequesterProvider =
+        BinaryProcessRequesterProvider.create(binaryRun, binaryProcessGatewayProvider, 0);
+
+    executeOnDead(10);
+    staticConfigMockedStatic.close();
+
+    verify(binaryProcessGateway, timesBeyondCreation(10)).init(any());
+  }
+
+  @Test
+  public void shouldResetRestartAttemptCounterOnSuccess() {
+    when(mockedFuture.isDone()).thenReturn(true);
+    binaryProcessRequesterProvider =
+        BinaryProcessRequesterProvider.create(binaryRun, binaryProcessGatewayProvider, 0);
+
+    MockedStatic<StaticConfig> staticConfigMockedStatic = mockExponentialBackoffWith(0);
+    for (int i = 0; i < 10; i++) {
+      executeOnDead();
+      int restartAttemptCounter = i;
+      staticConfigMockedStatic.verify(
+          () -> StaticConfig.exponentialBackoff(Mockito.eq(restartAttemptCounter)), times(1));
+    }
+    staticConfigMockedStatic.close();
```

Comment:
why do you need to close and re-open this mock?

Reply:
because I want to run a verification on a clean mock and not take into account the previous calls (when running times(n))